packages/aws-lambda/src/triggers.js

```diff
@@ -212,43 +212,112 @@ function extractEventFromSQS(event, span) {
-        }
-      }
-    });
-  } else if (custom) {
-    if (custom['x-instana-l']) {
-      tracingHeaders.l = custom['x-instana-l'];
+    tracingHeaders.t = tracingUtil.readAttribCaseInsensitive(event.headers, tracingConstants.traceIdHeaderName);
+    tracingHeaders.s = tracingUtil.readAttribCaseInsensitive(event.headers, tracingConstants.spanIdHeaderName);
+    tracingHeaders.l = tracingUtil.readAttribCaseInsensitive(event.headers, tracingConstants.traceLevelHeaderName);
+    if (hasTracingHeaders(tracingHeaders)) {
+      return tracingHeaders;
+    }
+  }
+
+  const custom = readClientContextCustom(context);
+  if (custom) {
+    if (custom[tracingConstants.traceLevelHeaderNameLowerCase]) {
+      tracingHeaders.l = custom[tracingConstants.traceLevelHeaderNameLowerCase];
     }
-    if (custom['x-instana-s']) {
-      tracingHeaders.s = custom['x-instana-s'];
+    if (custom[tracingConstants.spanIdHeaderNameLowerCase]) {
+      tracingHeaders.s = custom[tracingConstants.spanIdHeaderNameLowerCase];
     }
-    if (custom['x-instana-t']) {
-      tracingHeaders.t = custom['x-instana-t'];
+    if (custom[tracingConstants.traceIdHeaderNameLowerCase]) {
+      tracingHeaders.t = custom[tracingConstants.traceIdHeaderNameLowerCase];
+    }
+    if (hasTracingHeaders(tracingHeaders)) {
+      return tracingHeaders;
+    }
+  }
+
+  if (isSQSTrigger(event)) {
+    const sqsMessageAttributes = event.Records[0].messageAttributes;
+    if (sqsMessageAttributes) {
+      readTracingAttributesFromSqs(sqsMessageAttributes, tracingHeaders);
+      if (hasTracingHeaders(tracingHeaders)) {
+        return tracingHeaders;
+      }
+    }
+
+    const sqsMessageBody = event.Records[0].body;
+    if (typeof sqsMessageBody === 'string' && sqsMessageBody.startsWith('{')) {
+      try {
+        const parsedSqsMessageBody = JSON.parse(sqsMessageBody);
+        const snsAttributes = parsedSqsMessageBody && parsedSqsMessageBody.MessageAttributes;
+        if (snsAttributes) {
+          readTracingAttributesFromSqs(snsAttributes, tracingHeaders);
+          if (hasTracingHeaders(tracingHeaders)) {
+            return tracingHeaders;
+          }
+        }
+      } catch (e) {
+        // The attempt to parse the message body as JSON failed, so this is not an SQS message resulting from an SNS
+        // notification (SNS-to-SQS subscription), in which case we are not interested in the body. Ignore the error and
+        // move on.
+      }
     }
   }
+
   return tracingHeaders;
 };
+
+function readClientContextCustom(context) {
+  // The Node.js AWS SDK documentation expects for Custom, with capital "C", but the same is not explicitly said for
+  // other languages, so we test both.
+  return (
+    (context && context.clientContext && context.clientContext.Custom) ||
+    (context && context.clientContext && context.clientContext.custom)
+  );
+}
+
+function readTracingAttributesFromSqs(attributes, tracingHeaders) {
```

Comment:
The way `readTracingHeaders` is implemented like is roughly this:

```
function readTracingHeaders(event, context) {
  step 1: try to read tracing headers from HTTP attributes;
  if (hasTracingHeaders(tracingHeaders)) {
    return tracingHeaders;
  }
  step 2: try to read tracing headers from ClientContext.Custom 
  if (hasTracingHeaders(tracingHeaders)) {
    return tracingHeaders;
  }
  step 3: try to read tracing headers from the SQS message attributes 
  if (hasTracingHeaders(tracingHeaders)) {
    return tracingHeaders;
  }
  step 4: try to read tracing headers from the SNS-to-SQS 
  
  return tracingHeaders;
}
```

So after each attempt, the function will check if it found any Instana tracing header/message attribute, and if so, exit early without attempting the other checks.

But you are right, this is not very straight forward and has potential for future maintenance issues because that pattern gets overlooked once we add the next trace correlation mechanisms. I couldn't come up with a better structure so far. 

I agree that this can probably be improved. Do you have a suggestion?

Reply:
All of the individual steps (steps 1-4) would mutate `tracingHeaders`. Maybe it would already be a small improvement if each of those steps created a brand new object instead. This would maybe make the intent clearer. The previous  `tracingHeaders` will be empty anyway, due the `hasTracingHeaders` checks. WDYT?

Reply:
Yes, I would probably go for returning brand new objects.
Since something similar was already in place, I would say: let's leave this as is and we tackle them in the upcoming refactoring engineering requests we already have.
If this was the only case just being introduced to the code, it would have been better to do it now, IMO.

Reply:
I've updated the whole `readTracingHeaders` thing (and renamed it to `readTraceCorrelationData`). WDYT, does that look better now?

Reply:
Yes. I think the returning immutable new object seems to be a better practice than having a void function that changes an existent object by reference.
Thanks for putting the effort into this refactoring now 👍🏽